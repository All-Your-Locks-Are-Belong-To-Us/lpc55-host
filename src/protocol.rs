// https://www.nxp.com/docs/en/reference-manual/MCUBOOTRM.pdf
//
// - all fields in packets are little-endian
// - each command sent from host is replied to with response
// - optional data phase, either command or response (not both!)
//   RM uses "incoming" (host->MCU) and "outgoing" (host<-MCU) terminology
//
// 1) no data phase:
//   -> command
//   <- ACK
//   <- generic response
//   -> ACK
//
// 2) command ("incoming") data phase:
//   -> command (has-data-phase flag set)
//   <- ACK
//   <- initial generic response
//   -> ACK
//   -> data
//   <- ACK
//   ...
//   -> final data
//   <- ACK
//   <- final generic response
//   -> ACK
//
// 3) response ("outgoing") data phase:
//   -> cmmand
//   <- ACK
//   <- initial response (has-data-phase flag set)
//   -> ACK
//   <- initial data packet
//   -> ACK
//   ...
//   <- final data packet
//   -> ACK

use hidapi::{HidDevice, HidResult};

pub struct Protocol {
    device: HidDevice,
}

impl Protocol {
    pub fn write(&self, data: &[u8]) -> HidResult<usize> {
        self.device.write(data)
    }

    pub fn read_timeout(&self, timeout: usize) -> HidResult<Vec<u8>> {
        let mut data = Vec::new();
        data.resize(256, 0);
        let read = self.device.read_timeout(&mut data, timeout as i32)?;
        data.resize(read, 0);
        Ok(data)
    }
}

impl Protocol {
    pub fn new(device: HidDevice) -> Self {
        Self { device }
    }
}

impl std::fmt::Debug for Protocol {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Device")
            // .debug_struct("HidDevice")
                .field("manufacturer", &self.device.get_manufacturer_string())
                .field("product", &self.device.get_product_string())
                .field("serial number", &self.device.get_serial_number_string())
            // .finish()
        .finish()
    }
}

