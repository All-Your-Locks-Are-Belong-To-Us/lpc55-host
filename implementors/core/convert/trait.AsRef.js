(function() {var implementors = {};
implementors["lpc55"] = [{"text":"impl AsRef&lt;[u8]&gt; for Keycode","synthetic":false,"types":[]},{"text":"impl AsRef&lt;[u8]&gt; for ActivationCode","synthetic":false,"types":[]},{"text":"impl AsRef&lt;[u8]&gt; for RawCustomerData","synthetic":false,"types":[]},{"text":"impl AsRef&lt;[u8]&gt; for Signature","synthetic":false,"types":[]},{"text":"impl AsRef&lt;[u8]&gt; for Sha256Hash","synthetic":false,"types":[]}];
if (window.register_implementors) {window.register_implementors(implementors);} else {window.pending_implementors = implementors;}})()