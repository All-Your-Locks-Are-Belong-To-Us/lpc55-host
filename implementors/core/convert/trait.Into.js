(function() {var implementors = {};
implementors["lpc55"] = [{"text":"impl Into&lt;(Group, u8)&gt; for BootloaderError","synthetic":false,"types":[]}];
if (window.register_implementors) {window.register_implementors(implementors);} else {window.pending_implementors = implementors;}})()