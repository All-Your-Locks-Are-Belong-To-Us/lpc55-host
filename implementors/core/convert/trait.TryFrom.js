(function() {var implementors = {};
implementors["lpc55"] = [{"text":"impl TryFrom&lt;u32&gt; for ErrorGroup","synthetic":false,"types":[]},{"text":"impl TryFrom&lt;u8&gt; for GenericError","synthetic":false,"types":[]},{"text":"impl TryFrom&lt;u8&gt; for FlashDriverError","synthetic":false,"types":[]},{"text":"impl TryFrom&lt;u8&gt; for SbLoaderError","synthetic":false,"types":[]},{"text":"impl TryFrom&lt;u8&gt; for PropertyStoreError","synthetic":false,"types":[]},{"text":"impl TryFrom&lt;u8&gt; for CrcCheckerError","synthetic":false,"types":[]},{"text":"impl TryFrom&lt;ReceivedPacket&gt; for ResponsePacket","synthetic":false,"types":[]},{"text":"impl TryFrom&lt;ReceivedPacket&gt; for Vec&lt;u8&gt;","synthetic":false,"types":[]},{"text":"impl&lt;'_&gt; TryFrom&lt;&amp;'_ [u8]&gt; for ProtectedFlash","synthetic":false,"types":[]},{"text":"impl&lt;'_&gt; TryFrom&lt;&amp;'_ [u8]&gt; for InfieldAreas","synthetic":false,"types":[]},{"text":"impl&lt;CustomerData, VendorUsage, '_&gt; TryFrom&lt;&amp;'_ [u8]&gt; for InfieldArea&lt;CustomerData, VendorUsage&gt; <span class=\"where fmt-newline\">where<br>&nbsp;&nbsp;&nbsp;&nbsp;CustomerData: InfieldAreaCustomerData,<br>&nbsp;&nbsp;&nbsp;&nbsp;VendorUsage: InfieldAreaVendorUsage,&nbsp;</span>","synthetic":false,"types":[]},{"text":"impl&lt;CustomerData, VendorUsage, '_&gt; TryFrom&lt;&amp;'_ [u8]&gt; for FactorySettings&lt;CustomerData, VendorUsage&gt; <span class=\"where fmt-newline\">where<br>&nbsp;&nbsp;&nbsp;&nbsp;CustomerData: FactorySettingsCustomerData,<br>&nbsp;&nbsp;&nbsp;&nbsp;VendorUsage: FactorySettingsVendorUsage,&nbsp;</span>","synthetic":false,"types":[]},{"text":"impl&lt;'_&gt; TryFrom&lt;&amp;'_ [u8]&gt; for Keystore","synthetic":false,"types":[]},{"text":"impl&lt;'a&gt; TryFrom&lt;&amp;'a BootCommandDescriptor&gt; for BootCommand","synthetic":false,"types":[]},{"text":"impl&lt;'_&gt; TryFrom&lt;&amp;'_ str&gt; for Config","synthetic":false,"types":[]},{"text":"impl&lt;'_&gt; TryFrom&lt;&amp;'_ str&gt; for SigningKeySource","synthetic":false,"types":[]},{"text":"impl&lt;'a&gt; TryFrom&lt;&amp;'a [u8]&gt; for Signature","synthetic":false,"types":[]},{"text":"impl TryFrom&lt;u8&gt; for FlashReadMargin","synthetic":false,"types":[]},{"text":"impl&lt;'_&gt; TryFrom&lt;&amp;'_ str&gt; for Key","synthetic":false,"types":[]},{"text":"impl TryFrom&lt;u8&gt; for ResponseTag","synthetic":false,"types":[]},{"text":"impl TryFrom&lt;u8&gt; for ReportId","synthetic":false,"types":[]}];
if (window.register_implementors) {window.register_implementors(implementors);} else {window.pending_implementors = implementors;}})()