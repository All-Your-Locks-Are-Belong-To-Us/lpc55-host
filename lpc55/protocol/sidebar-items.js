initSidebarItems({"constant":[["READ_TIMEOUT",""]],"enum":[["Error","The NXP bootloader protocol error type"],["ReceivedPacket",""]],"struct":[["Protocol","The NXP bootloader protocol. Interact via `fn call(Command) -> Result<Response>`"],["ResponsePacket",""]],"type":[["Result","The NXP bootloader protocol result types"]]});