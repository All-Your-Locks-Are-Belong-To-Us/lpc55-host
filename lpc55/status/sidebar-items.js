initSidebarItems({"enum":[["BootloaderError",""],["CrcCheckerError",""],["FlashDriverError",""],["GenericError",""],["Group",""],["PropertyStoreError",""]]});