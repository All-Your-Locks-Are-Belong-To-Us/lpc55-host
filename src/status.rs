use core::convert::TryFrom;

use serde::{Deserialize, Serialize};

#[repr(u8)]
#[derive(Clone, Debug, enum_iterator::IntoEnumIterator)]
pub enum Group {
    Generic = 0,
    FlashDriver = 1,
    QuadSpiDriver = 4,
    OtfadDriver = 5,
    Bootloader = 100,
    SbLoader = 101,
    MemoryInterface = 102,
    PropertyStore = 103,
    CrcChecker = 104,
    Packetizer = 105,
    ReliableUpdate = 106,
}

impl TryFrom<u32> for Group {
    type Error = u32;
    fn try_from(group: u32) -> core::result::Result<Self, u32> {
        use Group::*;
        Ok(match group {
            0 => Generic,
            1 => FlashDriver,
            4 => QuadSpiDriver,
            5 => OtfadDriver,
            100 => Bootloader,
            101 => SbLoader,
            102 => MemoryInterface,
            103 => PropertyStore,
            104 => CrcChecker,
            105 => Packetizer,
            106 => ReliableUpdate,
            _ => return Err(group),
        })
    }
}

#[derive(Copy, Clone, Debug, Deserialize, PartialEq, Serialize)]
pub enum BootloaderError {
    Generic(GenericError),
    FlashDriver(FlashDriverError),
    PropertyStore(PropertyStoreError),
    CrcChecker(CrcCheckerError),
    Unknown(u32),
}

#[repr(u8)]
#[derive(Copy, Clone, Debug, Deserialize, enum_iterator::IntoEnumIterator, PartialEq, Serialize)]
pub enum GenericError {
    Fail = 1,
    ReadOnly = 2,
    OutOfRange = 3,
    InvalidArgument = 4,
    Timeout = 5,
    NoTransferInProgress = 6,
}

impl TryFrom<u8> for GenericError {
    type Error = u8;
    fn try_from(code: u8) -> core::result::Result<Self, u8> {
        use GenericError::*;
        Ok(match code {
            1 => Fail,
            2 => ReadOnly,
            3 => OutOfRange,
            4 => InvalidArgument,
            5 => Timeout,
            6 => NoTransferInProgress,
            _ => return Err(code),
        })
    }
}

#[repr(u8)]
#[derive(Copy, Clone, Debug, Deserialize, enum_iterator::IntoEnumIterator, PartialEq, Serialize)]
pub enum FlashDriverError {
    Size = 0,
    Alignment = 1,
    Address = 2,
    Access = 3,
    ProtectionViolation = 4,
    CommandFailure = 5,
    UnknownProperty = 6,
    EraseKeyError = 7,
    ExecuteOnlyRegion = 8,
    UnsupportedApi = 15,
}

impl TryFrom<u8> for FlashDriverError {
    type Error = u8;
    fn try_from(code: u8) -> core::result::Result<Self, u8> {
        use FlashDriverError::*;
        Ok(match code {
            0 => Size,
            1 => Alignment,
            2 => Address,
            3 => Access,
            4 => ProtectionViolation,
            5 => CommandFailure,
            6 => UnknownProperty,
            7 => EraseKeyError,
            8 => ExecuteOnlyRegion,
            15 => UnsupportedApi,
            _ => return Err(code),
        })
    }
}

#[repr(u8)]
#[derive(Copy, Clone, Debug, Deserialize, enum_iterator::IntoEnumIterator, PartialEq, Serialize)]
pub enum PropertyStoreError {
    UnknownProperty = 0,
    ReadOnlyProperty = 1,
    InvalidValue = 2,
}

impl TryFrom<u8> for PropertyStoreError {
    type Error = u8;
    fn try_from(code: u8) -> core::result::Result<Self, u8> {
        use PropertyStoreError::*;
        Ok(match code {
            0 => UnknownProperty,
            1 => ReadOnlyProperty,
            2 => InvalidValue,
            _ => return Err(code),
        })
    }
}

#[repr(u8)]
#[derive(Copy, Clone, Debug, Deserialize, enum_iterator::IntoEnumIterator, PartialEq, Serialize)]
pub enum CrcCheckerError {
    Passed = 0,
    Failed = 1,
    Inactive = 2,
    Invalid = 3,
    OutOfRange = 4,
}

impl TryFrom<u8> for CrcCheckerError {
    type Error = u8;
    fn try_from(code: u8) -> core::result::Result<Self, u8> {
        use CrcCheckerError::*;
        Ok(match code {
            0 => Passed,
            1 => Failed,
            2 => Inactive,
            3 => Invalid,
            4 => OutOfRange,
            _ => return Err(code),
        })
    }
}

impl Into<(Group, u8)> for BootloaderError {
    fn into(self) -> (Group, u8) {
        use BootloaderError::*;
        match self {
            Generic(error) => (Group::Generic, error as u8),
            FlashDriver(error) => (Group::FlashDriver, error as u8),
            PropertyStore(error) => (Group::PropertyStore, error as u8),
            CrcChecker(error) => (Group::CrcChecker, error as u8),
            Unknown(_status) => panic!(),
        }
    }
}

impl From<BootloaderError> for u32 {
    fn from(error: BootloaderError) -> u32 {
        let (group, code) = error.into();
        (group as u32 * 100) + code as u32
    }
}

impl From<u32> for BootloaderError {
    fn from(status: u32) -> Self {
        use BootloaderError::*;
        if let Ok(group) = Group::try_from(status/100) {
            let code = (status % 100) as u8;
            match (group, code) {
                (Group::Generic, code) => GenericError::try_from(code).map_or(Unknown(status), Generic),
                (Group::FlashDriver, code) => FlashDriverError::try_from(code).map_or(Unknown(status), FlashDriver),
                (Group::PropertyStore, code) => PropertyStoreError::try_from(code).map_or(Unknown(status), PropertyStore),
                (Group::CrcChecker, code) => CrcCheckerError::try_from(code).map_or(Unknown(status), CrcChecker),
                _ => return Unknown(status),
            }
        } else {
            Unknown(status)
        }
    }
}
