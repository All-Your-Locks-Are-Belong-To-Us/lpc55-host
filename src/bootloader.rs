use core::convert::TryInto;

use hidapi::HidApi;
use enum_iterator::IntoEnumIterator;

use crate::protocol::Protocol;
use crate::status::BootloaderError;
use crate::types::*;
use crate::Result;

#[derive(Debug)]
pub struct Bootloader {
    protocol: Protocol,
    // move around; also "new" should scan the device_list iterator
    // to pull out all the info
    vid: u16,
    pid: u16,
}

impl Bootloader {
    pub fn try_new(vid: u16, pid: u16) -> Result<Self> {
        let api = HidApi::new()?;
        let device = api.open(vid, pid)?;
        let protocol = Protocol::new(device);
        Ok(Self { protocol, vid, pid } )
    }

    // pub fn command(&self, command: Command) {
    // }

    pub fn info(&self) {
        for property in Property::into_enum_iter() {
            println!("\n{:?}", property);
            self.property(property).ok();
        }
    }

    /// TODO: Need to expect erorrs such as: `Response status = 139 (0x8b) kStatus_FLASH_NmpaUpdateNotAllowed`
    /// This happens with `read-memory $((0x0009_FC70)) 16`, which would be the UUID
    ///
    /// TODO: Need to expect errors such as: `Response status = 10200 (0x27d8) kStatusMemoryRangeInvalid`
    /// This happens with `read-memory $((0x5000_0FFC)) 1`, which would be the DIEID (for chip rev)
    pub fn read_memory(&self, address: usize, length: usize) -> Vec<u8> {
        let mut data = Vec::new();
        let mut remaining = length;
        let mut address = address;
        while remaining > 0 {
            let length = core::cmp::min(remaining, 512);
            data.extend_from_slice(&self.read_memory_at_most_512(address, length));
            remaining -= length;
            address += length;
        }
        data
    }

    pub fn read_memory_at_most_512(&self, address: usize, length: usize) -> Vec<u8> {

        // construct
        let packet = Command::ReadMemory { address, length }.hid_packet();

        // send
        // println!("send ({}): {}", packet.len(), to_hex_string(&packet));
        let _wrote = self.protocol.write(packet.as_slice()).expect("could not write");
        //assert_eq!(_wrote, packet.len());

        // expected memory readout from `blhost`:
        // 00 00 04 20 c7 56 00 00 0b 57 00 00 5d 8d 00 00
        // 0b 57 00 00 0b 57 00 00 0b 57 00 00 0b 57 00 00
        // 00 00 00 00 00 00 00 00 00 00 00 00 0b 57 00 00
        // 0b 57 00 00 00 00 00 00 0b 57 00 00 0b 57 00 00

        // what happens:
        // - we send
        //                         Read    n_parms   addr     amount 512
        // send (36): 01 00 20 00  03 00 00 02  00 00 00 00  00 02 00 00  00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00
        //                0xC = 12 = header + 2 params                                                                                     Junk ->
        // read (60): 03 00 0C 00  A3 01 00 02  00 00 00 00  00 02 00 00  00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00  FA AB D2 99 02 F0 23 F9 FF E7 F4 A8 05 F0 05 FC CF 90 FF E7 CF 98 00 F0
        //
        // read (60): 04 00 38 00  00 00 04 20 C7 56 00 00 0B 57 00 00 5D 8D 00 00 0B 57 00 00 0B 57 00 00 0B 57 00 00 0B 57 00 00 00 00 00 00 00 00 00 00 00 00 00 00 0B 57 00 00 0B 57 00 00 00 00 00 00
        // read (60): 04 00 38 00  0B 57 00 00 0B 57 00 00 0B 57 00 00 0B 57 00 00 0B 57 00 00 0B 57 00 00 0B 57 00 00 0B 57 00 00 0B 57 00 00 0B 57 00 00 0B 57 00 00 0B 57 00 00 0B 57 00 00 0B 57 00 00
        // read (60): 04 00 38 00  0B 57 00 00 0B 57 00 00 0B 57 00 00 0B 57 00 00 0B 57 00 00 0B 57 00 00 0B 57 00 00 0B 57 00 00 0B 57 00 00 0B 57 00 00 0B 57 00 00 00 00 00 00 00 00 00 00 00 00 00 00
        // read (60): 04 00 38 00  00 00 00 00 0B 57 00 00 0B 57 00 00 0B 57 00 00 00 00 00 00 0B 57 00 00 0B 57 00 00 0B 57 00 00 0B 57 00 00 0B 57 00 00 0B 57 00 00 0B 57 00 00 0B 57 00 00 00 00 00 00
        // read (60): 04 00 38 00  00 00 00 00 00 00 00 00 0B 57 00 00 00 00 00 00 00 00 00 00 00 00 00 00 0B 57 00 00 0B 57 00 00 0B 57 00 00 00 00 00 00 0B 57 00 00 0B 57 00 00 0B 57 00 00 00 00 00 00
        // read (60): 04 00 38 00  0B 57 00 00 00 00 00 00 0B 57 00 00 00 00 00 00 0B 57 00 00 0B 57 00 00 80 B5 6F 46 00 F0 01 F8 FE DE B0 B5 02 AF AD F5 06 5D 0D F5 A3 50 D5 A9 D4 90 08 46 02 F0 8C F9
        // read (60): 04 00 38 00  FF E7 F0 A8 01 F0 59 FF FF E7 60 20 03 F0 C1 FB D3 90 FF E7 EC A8 F0 A9 D3 9A 01 F0 68 FF FF E7 68 46 E6 A9 01 60 E9 A8 EC A9 E4 AA E5 AB 04 F0 D3 FD FF E7 48 F6 9C 51
        // read (60): 04 00 38 00  C0 F2 00 01 E9 A8 02 F0 3E F8 8D F8 A0 13 E7 90 FF E7 48 F6 C0 50 C0 F2 00 00 01 68 48 F6 D8 50 C0 F2 00 00 00 68 FC 90 FC 98 D4 9A C2 F8 1C 0C 42 F2 95 03 C0 F2 00 03
        // read (60): 04 00 38 00  D2 91 19 46 05 F0 00 F9 D1 90 D0 91 FF E7 D1 98 FA 90 D0 99 FB 91 6A 46 01 23 13 60 F4 A8 02 22 FA AB D2 99 02 F0 23 F9 FF E7 F4 A8 05 F0 05 FC CF 90 FF E7 CF 98 00 F0
        // read (60): 04 00 08 00  01 00 48 F6 DC 51 C0 F2 D1 90 D0 91 FF E7 D1 98 FA 90 D0 99 FB 91 6A 46 01 23 13 60 F4 A8 02 22 FA AB D2 99 02 F0 23 F9 FF E7 F4 A8 05 F0 05 FC CF 90 FF E7 CF 98 00 F0
        //
        // read (60): 03 00 0C 00  A0 00 00 02  00 00 00 00  03 00 00 00  FF E7 D1 98 FA 90 D0 99 FB 91 6A 46 01 23 13 60 F4 A8 02 22 FA AB D2 99 02 F0 23 F9 FF E7 F4 A8 05 F0 05 FC CF 90 FF E7 CF 98 00 F0
        // read (0):

        // fetch response
        let initial_generic_response = self.protocol.read_timeout(2000).expect("generic reponse first");
        // dbg!(to_hex_string(&initial_generic_response));
        assert_eq!([0x03, 0x00, 0x0C, 0x00], &initial_generic_response[..4]);
        assert_eq!([0xA3, 0x01, 0x00, 0x02], &initial_generic_response[4..][..4]);

        let mut remaining = length;
        let mut data = Vec::new();
        while remaining > 0 {
            // dbg!(remaining);
            let response = self.protocol.read_timeout(2000).expect("could not read");
            assert!(response.len() >= 5);
            // dbg!(to_hex_string(&response));
            // println!("read ({}): {}", response.len(), to_hex_string(&response));
            let [report_id, _, count, _]: [u8; 4] = response.as_slice()[..4].try_into().unwrap();
            assert_eq!(report_id, ReportId::DataIn as u8);
            let count = count as usize;
            assert!(response.len() >= 4 + count);
            assert!(count as usize <= remaining);
            data.extend_from_slice(&response[4..][..count]);
            remaining -= count;
        }

        let final_generic_response = self.protocol.read_timeout(2000).expect("generic reponse first");
        assert_eq!([0x03, 0x00, 0x0C, 0x00], &final_generic_response[..4]);
        assert_eq!([0xA0, 0x00, 0x00, 0x02], &final_generic_response[4..][..4]);

        data
    }

    // INFO:MBOOT:CMD: GetProperty(CurrentVersion, index=0)
    // DEBUG:MBOOT:TX-PACKET: Tag=GetProperty, Flags=0x00, P[0]=0x00000001, P[1]=0x00000000
    // DEBUG:MBOOT:USB:report-id == 1
    // DEBUG:MBOOT:USB:HID-HEADER[4]: 01 00 20 00
    // DEBUG:MBOOT:USB:OUT[59]: 01 00 20 00 07 00 00 02 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
    // writing 01 00 20 00 07 00 00 02 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00, (index: 32
    // DEBUG:MBOOT:USB:IN [59]: 03 00 0C 00 A7 00 00 02 00 00 00 00 00 00 03 4B 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
    // DEBUG:MBOOT:RX-PACKET: Tag=GetPropertyResponse, Status=Success, V[0]=0x4B030000
    // INFO:MBOOT:CMD: Done successfully

    fn property(&self, property: Property) -> std::result::Result<Vec<u32>, BootloaderError> {

        // construct command
        let packet = Command::GetProperty(property).hid_packet();

        // send command
        // println!("writing {:02x?}", &packet);
        let _wrote = self.protocol.write(packet.as_slice()).expect("could not write");
        // assert_eq!(_wrote, packet.len());

        // fetch response
        let response = self.protocol.read_timeout(2000).expect("could not read");
        // println!("read: {:?}", &response[..read]);

        // interpret response
        let (hid_header, rest) = response.split_at(4);
        assert_eq!(hid_header[..2], (ReportId::CommandIn as u8 as u16).to_le_bytes());

        let (packet_header, rest) = rest.split_at(4);
        assert_eq!(packet_header[..3], [Response::GetProperty as u8, 0, 0]);
        let num_params = packet_header[3];
        assert!(num_params >= 1);

        // status
        let (status, rest) = rest.split_at(4);
        let status = u32::from_le_bytes(status.as_ref().try_into().unwrap());
        if status != 0 {
            return Err(BootloaderError::from(status));
        }

        let mut values = Vec::new();
        for value_bytes in rest.chunks_exact(4).take(num_params as usize - 1) {
            let value = u32::from_le_bytes(value_bytes.as_ref().try_into().unwrap());
            values.push(value);
            // println!("value: {:02x?} (0x{:08x} = {})", value_bytes, value, value);
        }

        Ok(values)
    }

    pub fn properties(&self) -> GetProperties<'_> {
        GetProperties { bl: self }
    }

    pub fn all_properties(&self) -> Properties {
        let proxy = self.properties();
        Properties {
            current_version: proxy.current_version().unwrap(),
            target_version: proxy.target_version().unwrap(),
            available_commands: proxy.available_commands().unwrap(),
            available_peripherals: proxy.available_peripherals().unwrap(),
            pfr_keystore_update_option: proxy.pfr_keystore_update_option().unwrap(),
            ram_start_address: proxy.ram_start_address().unwrap(),
            ram_size: proxy.ram_size().unwrap(),
            flash_start_address: proxy.flash_start_address().unwrap(),
            flash_size: proxy.flash_size().unwrap(),
            flash_page_size: proxy.flash_page_size().unwrap(),
            flash_sector_size: proxy.flash_sector_size().unwrap(),
            verify_writes: proxy.verify_writes().unwrap(),
            flash_locked: proxy.flash_locked().unwrap(),
            max_packet_size: proxy.max_packet_size().unwrap(),
            device_uuid: proxy.device_uuid().unwrap(),
            system_uuid: proxy.system_uuid().unwrap(),
            crc_check_status: proxy.crc_check_status().unwrap(),
            reserved_regions: proxy.reserved_regions().unwrap(),
            irq_notification_pin: proxy.irq_notification_pin().unwrap(),
        }
    }
}

pub struct GetProperties<'a> {
    bl: &'a Bootloader
}

impl GetProperties<'_> {
    pub fn current_version(&self) -> std::result::Result<Version, BootloaderError> {
        Ok(Version::from(self.bl.property(Property::CurrentVersion)?[0]))
    }
    pub fn target_version(&self) -> std::result::Result<Version, BootloaderError> {
        Ok(Version::from(self.bl.property(Property::TargetVersion)?[0]))
    }
    pub fn ram_start_address(&self) -> std::result::Result<usize, BootloaderError> {
        Ok(self.bl.property(Property::RamStartAddress)?[0] as _)
    }
    pub fn ram_size(&self) -> std::result::Result<usize, BootloaderError> {
        Ok(self.bl.property(Property::RamSize)?[0] as _)
    }
    pub fn flash_start_address(&self) -> std::result::Result<usize, BootloaderError> {
        Ok(self.bl.property(Property::FlashStartAddress)?[0] as _)
    }
    pub fn flash_size(&self) -> std::result::Result<usize, BootloaderError> {
        Ok(self.bl.property(Property::FlashSize)?[0] as _)
    }
    pub fn flash_page_size(&self) -> std::result::Result<usize, BootloaderError> {
        Ok(self.bl.property(Property::FlashPageSize)?[0] as _)
    }
    pub fn flash_sector_size(&self) -> std::result::Result<usize, BootloaderError> {
        Ok(self.bl.property(Property::FlashSectorSize)?[0] as _)
    }
    pub fn max_packet_size(&self) -> std::result::Result<usize, BootloaderError> {
        Ok(self.bl.property(Property::MaxPacketSize)?[0] as _)
    }
    pub fn available_peripherals(&self) -> std::result::Result<crate::types::AvailablePeripherals, BootloaderError> {
        Ok(AvailablePeripherals::from_bits_truncate(self.bl.property(Property::AvailablePeripherals)?[0]))
    }
    pub fn available_commands(&self) -> std::result::Result<crate::types::AvailableCommands, BootloaderError> {
        Ok(AvailableCommands::from_bits_truncate(self.bl.property(Property::AvailableCommands)?[0]))
    }
    pub fn pfr_keystore_update_option(&self) -> std::result::Result<crate::types::PfrKeystoreUpdateOptions, BootloaderError> {
        let values = self.bl.property(Property::PfrKeystoreUpdateOptions)?;
        assert_eq!(values.len(), 1);
        Ok(PfrKeystoreUpdateOptions::from(values[0]))
    }
    pub fn verify_writes(&self) -> std::result::Result<bool, BootloaderError> {
        Ok(self.bl.property(Property::VerifyWrites)?[0] == 1)
    }
    pub fn flash_locked(&self) -> std::result::Result<bool, BootloaderError> {
        Ok(match self.bl.property(Property::FlashSecurityState)?[0] {
            0x0 | 0x5AA55AA5 => false,
            0x1 | 0xC33CC33C => true,
            _ => panic!(),
        })
    }
    pub fn device_uuid(&self) -> std::result::Result<u128, BootloaderError> {
        let values = self.bl.property(Property::UniqueDeviceIdent)?;
        assert_eq!(values.len(), 4);
        Ok(
            ((values[3] as u128) << 96) +
            ((values[2] as u128) << 64) +
            ((values[1] as u128) << 32) +
            ((values[0] as u128))
        )
    }
    pub fn system_uuid(&self) -> std::result::Result<u64, BootloaderError> {
        let values = self.bl.property(Property::SystemDeviceIdent)?;
        assert_eq!(values.len(), 2);
        Ok(((values[1] as u64) << 32) + (values[0] as u64))
    }

    pub fn crc_check_status(&self) -> std::result::Result<BootloaderError, BootloaderError> {
        Ok(BootloaderError::from(self.bl.property(Property::CrcCheckStatus)?[0]))
    }

    pub fn reserved_regions(&self) -> std::result::Result<Vec<(usize, usize)>, BootloaderError> {
        let values = self.bl.property(Property::ReservedRegions)?;
        assert_eq!(values.len() % 2, 0);
        let mut pairs = Vec::new();
        for pair in values.chunks_exact(2) {
            let left = pair[0];
            let right = pair[1];
            assert!(right >= left);
            if right > left {
                pairs.push((left as usize, right as usize));
            }
        }
        Ok(pairs)
    }

    pub fn irq_notification_pin(&self) -> std::result::Result<crate::types::IrqNotificationPin, BootloaderError> {
        Ok(IrqNotificationPin::from(self.bl.property(Property::IrqNotificationPin)?[0]))
    }
}

#[cfg(test)]
#[test]
fn test_all_properties() {
    let (vid, pid) = (0x1fc9, 0x0021);
    let bootloader = Bootloader::try_new(vid, pid).unwrap();
    insta::assert_debug_snapshot!(bootloader.all_properties());
}
