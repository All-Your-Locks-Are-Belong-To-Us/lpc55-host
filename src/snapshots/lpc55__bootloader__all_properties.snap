---
source: src/bootloader.rs
expression: bootloader.all_properties()
---
Properties {
    current_version: Version {
        mark: Some(
            'K',
        ),
        major: 3,
        minor: 0,
        fixation: 0,
    },
    target_version: Version {
        mark: Some(
            'T',
        ),
        major: 1,
        minor: 1,
        fixation: 4,
    },
    available_commands: FLASH_ERASE_ALL | FLASH_ERASE_REGION | READ_MEMORY | FLASH_SECURITY_DISABLE | GET_PROPERTY | RECEIVE_SB_FILE | CALL | RESET | FLASH_READ_RESOURCE,
    available_peripherals: UART | I2C | SPI | USB_HID,
    pfr_keystore_update_option: Keystore,
    ram_start_address: 536870912,
    ram_size: 262144,
    flash_start_address: 0,
    flash_size: 646656,
    flash_page_size: 512,
    flash_sector_size: 32768,
    verify_writes: true,
    flash_locked: false,
    max_packet_size: 56,
    device_uuid: 156090513919447561107476615196208161407,
    system_uuid: 1168442901135557,
    crc_check_status: CrcChecker(
        Invalid,
    ),
    reserved_regions: [
        (
            335544320,
            335568895,
        ),
        (
            67108864,
            67141631,
        ),
        (
            805306368,
            805330943,
        ),
        (
            536870912,
            536895487,
        ),
    ],
    irq_notification_pin: IrqNotificationPin {
        pin: 0,
        port: 0,
        enabled: false,
    },
}
